### ISSUE TITLE
doom-modeline on its own package

### ISSUE THREAD

**Issue by Townk** [hlissner/doom-emacs#136]
I know you mentioned in the README file about this, but is there a plan to get doom-modeline on its own? I had to pick up too much things from core-ui, core-lib and memoize to make it work stand alone.

**Comment by hlissner** [comment]
It is planned, but I haven't figured out how exactly I want to go about it. My modeline config has many idiosyncrasies I'd like to generalize, but I'll have a look at it next weekend. Think it's about time it happened.

**Comment by wpcarro** [comment]
@Townk do you mind sharing your version? I'm trying to disentangle the configuration, but after following the readme, I'm still experiencing bootstrapping issues. Realizing that calls to `add-hook!` need to be replaced as well since they're defined within the broader Doom ecosystem. But I do not expect this to be the only necessary change.

**Comment by hlissner** [comment]
@wpcarro `def-modeline-segment!` and `def-modeline!` calls too. Looks like the readme needs more work. In any case, another user, [torgeir](https://github.com/torgeir), managed to extract my modeline: https://github.com/torgeir/.emacs.d/blob/1f2c1894655053a680bfd8b9441062c1c3f6b525/site-lisp/t-doom-modeline/t-doom-modeline.el. Perhaps that can be of help?

**Comment by wpcarro** [comment]
Thanks for the reference, @hlissner. I am enjoying your work quite a bit, so I hope your energy and dedication remains high. One question I've had as I've been reading some of doom's source is the `function-name!` notation. Is the `!` intended to denote a macro? I'm a lisp novice, so forgive my ignorance if this is a well known convention.

**Comment by hlissner** [comment]
@wpcarro I'm glad you like it; I expect to be working on it for a very long time!

> Is the ! intended to denote a macro?

It indeed denotes a macro, but it is not a standard Lisp or Elisp convention.

I list all the naming conventions I use in Doom at the top of [core/core.el](https://github.com/hlissner/.emacs.d/blob/master/core/core.el#L5-L13) and what they mean. Only the first two are Lisp conventions, the rest are my own.

**Comment by torgeir** [comment]
[Seconded](https://github.com/hlissner/.emacs.d/issues/136#issuecomment-312460111)! I [updated with the latest stuff](https://github.com/torgeir/.emacs.d/commit/fb4b4bd4e37ae3da33216e02dfcde1076d93fd59) just yesterday. Had to byte compile it to get rid of some wierd `(void-function dir-path)` and `(void-function file-path)` errors from [these lines](https://github.com/hlissner/.emacs.d/blob/master/modules/ui/doom-modeline/config.el#L296-L303), but other that that its been working great for me for a long time.

**Comment by seagle0128** [comment]
I vote this proposal. Eager to get `doom-modeline` on mepla.
Suggest to make it light and fast.

**Comment by seagle0128** [comment]
This issue has been raised over one year ago. I am implementing myself on https://github.com/seagle0128/doom-modeline. Will submit to mepla once it's ready.

**Comment by seagle0128** [comment]
Already submitted on https://github.com/melpa/melpa/pull/5550.

**Comment by hlissner** [comment]
@seagle0128 Thank you for this! Sorry for never getting around to it. Once I've merged develop here, I'll be happy to help you out.

A few things I've noticed: you've copied a lot of code verbatim from Doom. I have no problem with that, but I suspect the folks at MELPA will have issue with the naming conventions. Also, I notice you are using the Doom modeline from the `master` branch. I'd highly recommend using the one on `develop` instead. Many bugs have been fixed.

**Comment by seagle0128** [comment]
@hlissner , Thanks for the comments! I will refactor and update the codes from `develop` branch later. It just works for now and need a lot of work to do, I think. You help should be great!

**Comment by seagle0128** [comment]
While I am porting new codes from `develop` branch, I found some issues.

1. Must `(require 'map)` which means requires emacs 25+. I'd like support 24.4+.
1. The icons of vcs and flycheck are incorrect. Not sure the root cause :(
1. Should check if `face` is nil in doom-modeline--make-xpm.

@hlissner, Could you please help to check these issue?  The codes are [here](https://github.com/seagle0128/doom-modeline/tree/develop)

**Comment by seagle0128** [comment]
> vcs status not changed after switching the branch

My solutions is https://github.com/seagle0128/doom-modeline/commit/10ec55adcc69a08cbc3df814280f5e47f5f043c3:
```elisp
(declare-function magit-toplevel "magit-git")
(defun doom-modeline-magit-post-refresh ()
  "Update vcs state in mode-line after refreshing in magit."
  (dolist (buf (buffer-list))
    (when (and (not (buffer-modified-p buf))
               (buffer-file-name buf)
               (file-exists-p (buffer-file-name buf))
               (file-in-directory-p (buffer-file-name buf) (magit-toplevel)))
      (with-current-buffer buf
        (vc-refresh-state)
        (doom-modeline--update-vcs)))))
(add-hook 'magit-post-refresh-hook #'doom-modeline-magit-post-refresh)
```

**Comment by seagle0128** [comment]
The PR https://github.com/melpa/melpa/pull/5550 has been merged.
https://github.com/seagle0128/doom-modeline is available on MEPLA now.

**Comment by torgeir** [comment]
And it works like a charm! 👏

**Comment by seagle0128** [comment]
I just released v0.8.1 of [doom-modeline](https://github.com/seagle0128/doom-modeline) package.
I rewrote `doom-modeline--buffer-file-name` and added 3 new styles: 

```elisp
Given ~/Projects/FOSS/emacs/lisp/comint.el
  truncate-from-project => ~/Projects/FOSS/emacs/l/comint.el
  truncate-with-project => emacs/l/comint.el
  truncate-except-project => ~/P/F/emacs/l/comint.el
```
The default style is still `truncate-upto-project`. 
In  my testing, at least **5x** faster than original one.
It's really deserved to try!
